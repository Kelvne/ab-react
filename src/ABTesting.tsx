import * as React from 'react';
import { ReactCookieProps, withCookies } from 'react-cookie';

interface IABTestingProps extends ReactCookieProps {
  cookieName: string;
}

interface IABTestingState {
  selected: number;
}

export class ABTesting extends React.Component<IABTestingProps, IABTestingState> {
  constructor(props: React.Props<ABTesting> & IABTestingProps) {
    super(props);
    const { children, cookies, cookieName } = props;

    const cookie = cookies
      ? cookies.get(cookieName)
      : null;

    if (cookie) {
      this.state = {
        selected: Number(cookie),
      };

      return;
    }

    const count = React.Children.count(children);
    const selected = Math.floor(Math.random() * count);

    this.state = {
      selected,
    };

    if (cookies) { cookies.set(cookieName, String(selected), { maxAge: 86400 * 7 }); }
  }

  public render() {
    const { children, cookies, cookieName } = this.props;
    const { selected } = this.state;

    const components = React.Children.toArray(children);

    if (cookies && components.length - 1 < selected) {
      cookies.remove(cookieName);
      return components[0];
    }

    return components[selected];
  }
}

export default withCookies(ABTesting) as React.ComponentType<IABTestingProps>;
