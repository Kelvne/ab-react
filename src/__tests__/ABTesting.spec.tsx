/* tslint:disable:no-var-requires */

import { render, shallow } from 'enzyme';
import * as React from 'react';

const Cookies = require('universal-cookie');

import { ABTesting } from '../ABTesting';

describe('ABTesting Component', () => {
  it('should render', () => {
    const component = shallow(<ABTesting cookieName='ab-testing' />);

    expect(component).toMatchSnapshot();
  });

  it('show render a child if one child', () => {
    const component = shallow(
      <ABTesting cookieName='ab-testing'>
        <div>Hello</div>
      </ABTesting>,
    );

    expect(component).toMatchSnapshot();
  });

  it('show set a cookie using the cookie name', () => {
    const cookies = new Cookies();
    cookies.set = jest.fn();

    const cookieName = 'ab-testing-cookie';

    shallow(
      <ABTesting cookies={cookies} cookieName={cookieName}>
        <div>Hi!</div>
      </ABTesting>,
    );

    expect(cookies.set).toBeCalledWith(cookieName, '0', { maxAge: 86400 * 7 });
  });

  it('show always render same version if cookie is set', () => {
    const cookies = new Cookies();
    cookies.get = jest.fn(() => '1');

    const cookieName = 'ab-testing-ok';

    const component = shallow(
      <ABTesting cookies={cookies} cookieName={cookieName}>
        <div>This should not be rendered.</div>
        <div>This should be rendered.</div>
      </ABTesting>,
    );

    expect(component).toMatchSnapshot();
  });
});
