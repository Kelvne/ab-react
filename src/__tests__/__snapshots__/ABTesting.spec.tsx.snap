// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABTesting Component should render 1`] = `""`;

exports[`ABTesting Component show always render same version if cookie is set 1`] = `
<div
  key=".1"
>
  This should be rendered.
</div>
`;

exports[`ABTesting Component show render a child if one child 1`] = `
<div
  key=".0"
>
  Hello
</div>
`;
